src/analysis/gMCS/GPR2models.m

```diff
@@ -0,0 +1,259 @@
+    end
+
+    % Include Exchange reactions
+    for gen = 1:length(genes)
+        model = addReaction(model,['DM_',genes{gen}],...
+            'metaboliteList',{genes{gen}},'stoichCoeffList',[1],...
+            'reversible',false,'printLevel',printLevel-1,...
+            'geneRule', genes{gen});
+    end
+
+    % converge isoforms (if neccesary)
+    if ~isempty(separate_isoform) && length(genes)>1
+        model = convergeIsoforms(model, separate_isoform, printLevel);
+    end
+
+    % Add objective function
+    model = addReaction(model,RXN,'reactionFormula',[RXN,' -> '],...
+        'printLevel',printLevel-1); % exchange reaction for objective metabolite
+    % Define objective function
+    model = changeObjective(model, RXN);
+
+    % Store the model
+    networks{i} = model;
+end
+
+% Calculate the number of genes per GPR rule
+rxnNumGenes = zeros(size(selected_rxns));
+for i = 1:length(selected_rxns)
+    rxn = selected_rxns(i);
+    idx_genes = find(metabolic_model.rxnGeneMat(rxn,:));
+    rxn_isoforms = metabolic_model.genes(idx_genes);
+    rxn_genes = unique(strtok(rxn_isoforms,separate_isoform));
+    rxnNumGenes(i) = length(rxn_genes);
+end
+
+end
+
+
+function model = modelParser(model,genes,head,node_parent,parent,printLevel)
+% This function look into the GPRparser and builds auxiliary nodes for each
+% layer of the GPR rules.
+% For example:
+%     g1 & g2 & g3 is one layer in the model,   RXN = g1 + g2 + g3
+%     (g1 | g2) & g3 is a two layer model:      RXN = (g1 | g2) + g3
+%                                               (g1 | g2) = g1
+%                                               (g1 | g2) = g2   
+
+
+
+%  Name nodes according to position and layer inside the GPR rule
+node = cell(size(head.children));
+for child=1:length(head.children)
+    node{child} = [node_parent,'_',num2str(child)];
+end
+model = addMetabolite(model, node);
+
+%  Expand model with every node, using OR/AND rules
+if strcmp(class(head),'OrNode')
+    for child=1:length(head.children)
+        model = addReaction(model,node{child},...
+            'reactionFormula',[node{child},' -> ',parent],'printLevel',printLevel-1);
+    end
+else
+    model = addReaction(model,strjoin(node,' & '),...
+        'reactionFormula',[strjoin(node,' + '),' -> ',parent],'printLevel',printLevel-1);
+end
+
+
+for i=1:length(head.children)
+    child = head.children(i);
+    % check if the child is final layer
+    if strcmp(class(child),'LiteralNode')
+        gen = genes(str2num(child.id));
+        model = addReaction(model,[gen{1},'_',node{i}],...
+            'metaboliteList',{gen{1},node{i}},'stoichCoeffList',[-1, 1],...
+            'printLevel',printLevel-1);
+    else
+        model = modelParser(model,genes,child,node{i},node{i},printLevel);
+    end
+end
+
+end
+
+function modelOut = reduceModel(model)
+% As the algorithm generates redundant nodes, this function is used to
```

Comment:
Wouldn't it be sufficient to activate checkDuplicate during the addReaction step? 
That way no duplicate reactions should be present.

Reply:
This function is created to remove unnecesary nodes that are generated in the process. The way the code is generated, each literal node is created as a "metabolite", and then they converge into the real demand node, the gene. This way I reduce the model to eliminate redundant nodes:
gene -> node1 -> rule1  == gen -> rule1
         -> node2 -> rule2                -> rule2

Reply:
I think I get it. Maybe convertToDNF() + reduce() on the head node would make sense here.
Like that all formulas would be disjunct normal forms. Admittedly I'm not sure, which version is faster.